lib/Transforms/SecretInsertMgmt/SecretInsertMgmtPatterns.h

```diff
@@ -43,12 +60,69 @@ struct ModReduceBefore : public OpRewritePattern<Op> {
                                 PatternRewriter &rewriter) const override;
 
  private:
-  bool isMul;
   bool includeFirstMul;
   Operation *top;
   DataFlowSolver *solver;
 };
 
+template <typename Op>
+struct MatchCrossLevel : public OpRewritePattern<Op> {
+  using OpRewritePattern<Op>::OpRewritePattern;
+
+  MatchCrossLevel(MLIRContext *context, int *scaleCounter, Operation *top,
+                  DataFlowSolver *solver)
+      : OpRewritePattern<Op>(context, /*benefit=*/1),
+        scaleCounter(scaleCounter),
+        top(top),
+        solver(solver) {}
+
+  LogicalResult matchAndRewrite(Op op,
+                                PatternRewriter &rewriter) const override;
+
+ private:
+  int *scaleCounter;
```

Comment:
Related to my earlier comment, a code comment here may be helpful to explain the role of scaleCounter.